keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
++    ContainerQoSResources qos_resources = 3;
+```
+
+#### RuntimeStatus
+
+Extend the `RuntimeStatus` message with new `resources` field that is used to
+communicate the available QoS-class resources from the runtime to the client.
+
+This information can be used by the client (kubelet) to validate QoS-class
+resource assignments before starting a pod. In future steps kubelet will patch
+this information into node status.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Information about the discovered resources
++    ResourcesInfo resources = 2;
++}
+
++// ResourcesInfo contains information about the resources discovered by the
++// runtime.
++message ResourcesInfo {
++    // Pod-level QoS resources available.
++    repeated QoSResourceInfo pod_qos_resources = 1;
++    // Container-level QoS resources available.
++    repeated QoSResourceInfo container_qos_resources = 2;
++}
+
++// QoSResourceInfo contains information about one type of QoS resource.
++message QoSResourceInfo {
++    // Name of the QoS resources.
++    string Name = 1;
++    // Mutable is set to true if this resource supports in-place updates i.e.
++    // the class of a running container or sandbox can be changed.
++    bool Mutable = 2;
++    // List of classes of this QoS resource.
++    repeated QoSResourceClassInfo classes = 3;
++}
+
++// QoSResourceClassInfo contains information about one class of certain
++// QoS resource.
++message QoSResourceClassInfo {
++    // Name of the class
++    string name = 1;
++    // Capacity is the number of maximum allowed simultaneous assignments into this class
++    // Zero means "infinite" capacity i.e. the usage is not restricted.
++    uint64 capacity = 2;
+ }
+```
+
+#### Consts
+
+Also, define "known" QoS-class resource types to more easily align container
+runtime implementations:
+
+```diff
++const (
++       // QoSResourceRdt is the name of the RDT QoS-class resource
++       QoSResourceRdt = "rdt"
++       // QoSResourceBlockio is the name of the blockio QoS-class resource
++       QoSResourceBlockio = "blockio"
++)
+```
+
+### Kubernetes API
+
+The PodSpec will be extended to support assignment of pod-level and
+container-level QoS-class devices. NodeStatus will be extended to include
+information about the available QoS-class resources on a node.
+
+#### PodSpec
+
+We introduce a new field, QoSResources into the existing ResourceRequirements
+struct. This will enable the assignment of QoS resources for containers.
+
+```diff
+ type ResourceRequirements struct {
+        // Limits describes the maximum amount of compute resources allowed.
+@@ -2195,6 +2198,10 @@ type ResourceRequirements struct {
+        // +featureGate=DynamicResourceAllocation
+        // +optional
+        Claims []ResourceClaim
++       // QoSResources specifies the QoS resources.
++       // +optional
++       QoSResources map[QoSResourceName]string
```

Comment:
I'm not loving this name, but we can argue that later.

Reply:
Naming is hard 🥸 I'm really open to suggestions. Just changed the naming recently (IMO in better direction) after feedback from @vinaykul

Reply:
names can evolve during impl